scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,386 @@
+import logging
+import sys
+import time
+from functools import partial
+from itertools import cycle
+from typing import Optional, List, Tuple, Union, Iterable
+
+import numpy as np
+import torch
+from torch import nn
+from tqdm import trange
+
+from scvi.dataset import GeneExpressionDataset
+from scvi.inference import Posterior
+from scvi.inference import Trainer
+from scvi.models.log_likelihood import compute_elbo
+
+logger = logging.getLogger(__name__)
+
+
+class JPosterior(Posterior):
+    def __init__(self, *args, mode=0, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mode = mode
+
+    def elbo(self, verbose: bool = False) -> float:
+        elbo = compute_elbo(self.model, self, mode=self.mode)
+        if verbose:
+            logger.info("ELBO : %.4f" % elbo)
+        return elbo
+
+
+class JVAETrainer(Trainer):
+    r"""
+    The VariationalInference class for the unsupervised training of an autoencoder.
+
+    :param model: A model instance from class ``JVAE``
+    :param discriminator: A model instance of a classifier (with logit output)
+    :param gene_dataset_list: list of gene_dataset instance like ``[CortexDataset(), SmfishDataset()]``
```

Comment:
does these dataset need to have a certain property ? (one included in the other ?)

Reply:
Not necessarily. The user has to provide a mapping of the genes from each dataset to `Union(d.gene_names for d in datasets)`

Reply:
great !